### PR TITLE
Add server side API interface 

### LINKED ISSUE #167: Introduce server side API interface
Currently only modules have an abstract interface for their custom API endpoints. If this can be extended to the transaction submission API endpoint we could mock out the HTTP server/client and build an mock API client/server pair that passes requests internally without network connections.

### PR DESCRIPTION
**Pull request by Maan2003**
resolve #167

### CODE REVIEW

**Comment by justinmoon** [comment]
Looks good to me.

**Comment by jkitman** [comment]
> resolve #167

Looks good!  However, do you think we could have the `ApiEndpoint` return the `base_name: Option<&str>`?  Might be a bit cleaner code / for mocking later.

I suppose the `MockApiEndpoint` implementation for testing should be done in a separate issue or should we wait on closing #167?

**Review by jkitman — CHANGES_REQUESTED** [review]
> resolve #167

Looks good!  However, do you think we could have the `ApiEndpoint` return the `base_name: Option<&str>`?  Might be a bit cleaner code / for mocking later.

I suppose the `MockApiEndpoint` implementation for testing should be done in a separate issue or should we wait on closing #167?

File: minimint/src/net/api.rs [hunk 1]
```diff
@@ -67,44 +84,57 @@ where
...
+            handler: |minimint, _params, transaction| {
+                let transaction: Transaction =
+                    match serde_json::from_slice(&serde_json::to_vec(&transaction).unwrap()) {
+                        Ok(t) => t,
+                        Err(_) => return Ok(http::Response::new(400)),
+                    };
+                let tx_id = transaction.tx_hash();
 
-    // TODO: give feedback in case of error
-    let body = Body::from_json(&tx_id).expect("encoding error");
```
Comment (jkitman) [inline minimint/src/net/api.rs:88]: Do errors in the request parsing kill the API thread?

Comment (Maan2003) [reply minimint/src/net/api.rs:88]: no, this errors in serialization of response into json, which should not happen

File: minimint/src/net/api.rs [hunk 2]
```diff
@@ -73,38 +90,50 @@ where
...
+                };
+                let tx_id = transaction.tx_hash();
 
-    // TODO: give feedback in case of error
-    let body = Body::from_json(&tx_id).expect("encoding error");
-    Ok(body.into())
-}
+                minimint
+                    .submit_transaction(transaction)
+                    .expect("Could not submit sign request to consensus");
```
Comment (elsirion) [inline minimint/src/net/api.rs:109]: Eventually we probably want to return errors instead of `expect`ing here.

**Review by elsirion — APPROVED** [review]

**Pull request merged**